x/crosschain/keeper/msg_server_refund_aborted_tx.go

```diff
@@ -0,0 +1,104 @@
+	}
+
+	// check if the cctx is aborted
+	if cctx.CctxStatus.Status != types.CctxStatus_Aborted {
+		return nil, errorsmod.Wrap(types.ErrInvalidStatus, "CCTX is not aborted")
+	}
+	// check if the cctx is not refunded
+	if cctx.CctxStatus.IsAbortRefunded {
+		return nil, errorsmod.Wrap(types.ErrUnableProcessRefund, "CCTX is already refunded")
+	}
+
+	// Check if aborted amount is available to maintain zeta accounting
+	if cctx.InboundTxParams.CoinType == common.CoinType_Zeta {
+		err := k.RemoveZetaAbortedAmount(ctx, GetAbortedAmount(cctx))
+		// if the zeta accounting is not found, it means the zeta accounting is not set yet and the refund should not be processed
+		if errors.Is(err, types.ErrUnableToFindZetaAccounting) {
+			return nil, errorsmod.Wrap(types.ErrUnableProcessRefund, err.Error())
+		}
+		// if the zeta accounting is found but the amount is insufficient, it means the refund can be processed but the zeta accounting is not maintained properly
+		if errors.Is(err, types.ErrInsufficientZetaAmount) {
+			ctx.Logger().Error("Zeta Accounting Error: ", err)
+		}
+	}
+
+	refundAddress, err := GetRefundAddress(cctx, msg.RefundAddress)
+	if err != nil {
+		return nil, errorsmod.Wrap(types.ErrInvalidAddress, err.Error())
+	}
+	// refund the amount
+	// use temporary context to avoid gas refunding issues and side effects
+	tmpCtx, commit := ctx.CacheContext()
+	err = k.RefundAbortedAmountOnZetaChain(tmpCtx, cctx, refundAddress)
+	if err != nil {
+		return nil, errorsmod.Wrap(types.ErrUnableProcessRefund, err.Error())
+	}
+	commit()
+
+	// set the cctx as refunded
+	cctx.CctxStatus.AbortRefunded(ctx.BlockTime().Unix())
+
+	k.SetCrossChainTx(ctx, cctx)
+
+	return &types.MsgRefundAbortedCCTXResponse{}, nil
+}
+
+// Set the proper refund address.
+// For BTC sender chain the refund address is the one provided in the message in the RefundAddress field.
+// For EVM chain with coin type ERC20 the refund address is the sender , but can be overridden by the RefundAddress field in the message.
+// For EVM chain with coin type Zeta the refund address is the tx origin, but can be overridden by the RefundAddress field in the message.
+// For EVM chain with coin type Gas the refund address is the tx origin, but can be overridden by the RefundAddress field in the message.
+
+func GetRefundAddress(cctx types.CrossChainTx, optionalRefundAddress string) (ethcommon.Address, error) {
+	// make sure a separate refund address is provided for a bitcoin chain as we cannot refund to tx origin or sender in this case
+	if common.IsBitcoinChain(cctx.InboundTxParams.SenderChainId) && optionalRefundAddress == "" {
+		return ethcommon.Address{}, errorsmod.Wrap(types.ErrInvalidAddress, "refund address is required for bitcoin chain")
+	}
+	refundAddress := ethcommon.HexToAddress(cctx.InboundTxParams.TxOrigin)
+	if cctx.InboundTxParams.CoinType == common.CoinType_ERC20 {
+		refundAddress = ethcommon.HexToAddress(cctx.InboundTxParams.Sender)
```

Comment:
why make an exception for ERC20 coin type? 

In general i think it would be hard to determine between tx.origin and msg.sender as refundAddress. (most of the time I think it most makes sense to refund to tx.origin as it's guaranteed to be an EOA, while msg.sender could be a contract address that is not controlled or deployed on zetachain). But this might also not be robust--imagine a user tx initiated by an account abstracted wallet in which EOA is not the user but some 3rd party. 

I'd recommend not defaulting to anything--the admin msg should supply an refundAddress.

Reply:
@brewmaster012 @lumtis 
Made the change , 

But please note, even though we can force a separate refund address to be mandatory here, in case of auto refunds, we always select msg. sender for ERC 20 (The existing functionality )  .

I don't see issues , but it might make sense to replicate the same logic everywhere for code simplicity. 

I am thinking of removing the auto refund unless we need it; in that case, all refunds can be processed similarly. 

Adding a refund address to the CCTX is also an option , but I don't think we should go that route and overcomplicate it.